### DISCUSSION TITLE
Python 3.6 update help

### DISCUSSION THREAD

**Discussion by EHRETic** [netbox-community/netbox#7039] [Q&A] [answered]
Hi there,

As you know, Netbox 3 will drop Python 3.6 support so I want to migrate to a newer version as soon as I can.

The thing is:
- I'm using CentOS 8 for the moment, planning to migrate "on the fly" to AlmaLinux, so I'll keep the RHEL8 family for a while.
- the default Python version on CentOS 8 is 3.6, I've no clue if this will change anytime soon and as I'm not a Linux pro, I installed the default
- I'd like to make this as clean as possible, so default would be let's say 3.9.X and I'd like to be updated automatically of course

Sound maybe stupid, but all article I could read would compile another Python aside the default version, but I don't know how this "works", so what will happen when default will be replaced, is the newly compiled version automatically updated, what happens if the new default is the one I've compiled, etc...

So anybody that can help and fill up the holes me is welcome! 😉
(And this guide might help a few others...)

**Comment by jeremystretch** [comment]
> Sound maybe stupid, but all article I could read would compile another Python aside the default version, but I don't know how this "works", so what will happen when default will be replaced, is the newly compiled version automatically updated, what happens if the new default is the one I've compiled, etc...

The _default_ Python installation is what the operating system uses natively, but you can install additional Python versions in parallel for use by individual applications. For example, I have 3.5 through 3.9 installed on my local system:

```
$ ls -l /usr/bin/python3.?
-rwxr-xr-x 2 root root 4117024 Feb 20  2021 /usr/bin/python3.5
-rwxr-xr-x 2 root root 4526456 Jan 26  2021 /usr/bin/python3.6
-rwxr-xr-x 2 root root 5308240 Feb 20  2021 /usr/bin/python3.7
-rwxr-xr-x 1 root root 5138128 Apr  2 21:00 /usr/bin/python3.8
-rwxr-xr-x 1 root root 5381008 Jul  3 13:50 /usr/bin/python3.9
```

The system's `python3` alias will link to `/usr/bin/python3.6`, but you can run any of these independently.

For NetBox v3.0, you can specify your desired Python version when executing the upgrade script using an environment variable, like this:

```
PYTHON=/usr/bin/python/3.7 upgrade.sh
```

**Comment by baldgeek** [comment, marked as answer]
I have a centos 8 netbox install that I moved to python 3.8 when I 
upgraded to 2.11.11 without issue using these commands. This also moved 
the system default to python 3.8

dnf install python38 (verify what is installing is what you want)

alternatives --config python3 (pick python3.8 from the list)

**Comment by EHRETic** [comment]
@baldgeek @jeremystretch thanks a lot for both your advice! 😉

I did try the Python update alone but Netbox didn't run anymore right after the update+reboot so as I also had a snapshot of my VM (NB 2.11.11), I reverted it so I would be able to run the upgrade script to 2.11.12 once again after the updated Python (3.9) which is also now the new default.

Is that OK? (As I didn't know if I can run the same upgrade script twice)

**Reply by candlerb** [reply]
Certainly it is safe to run `./upgrade.sh` multiple times.  Each time it will just recreate the virtualenv and check if there are any database migrations to apply (and if there are none, then it won't do any).

**Reply by EHRETic** [reply]
> Certainly it is safe to run `./upgrade.sh` multiple times. Each time it will just recreate the virtualenv and check if there are any database migrations to apply (and if there are none, then it won't do any).

Good to know, thanks a lot and it seems I migrated Python at the right time as 3.0 is out!!! 😁
